x/ibc/23-commitment/types/merkle.go

```diff
@@ -154,31 +183,292 @@ func (MerkleProof) GetCommitmentType() exported.Type {
-		return errors.New("empty params or proof")
+		return sdkerrors.Wrap(ErrInvalidProof, "empty params or proof")
 	}
 
-	runtime := rootmulti.DefaultProofRuntime()
-	return runtime.VerifyValue(proof.Proof, root.GetHash(), path.String(), value)
+	mpath, ok := path.(MerklePath)
+	if !ok {
+		return sdkerrors.Wrap(ErrInvalidProof, "path is not a merkle path for a merkle proof")
+	}
+
+	if len(proof.Proofs) != len(mpath.KeyPaths) || len(proof.Proofs) != len(specs) {
+		return sdkerrors.Wrapf(ErrInvalidProof, "invalid chained proof. chained proof length %d, spec length %d, path length %d must all be equal",
+			len(proof.Proofs), len(specs), len(mpath.KeyPaths))
+	}
+
+	for i, p := range proof.Proofs {
+		// While the proofs go from lowest subtree to the final tree, the path is defined from the root
+		// down to the leaf. Thus, we must pass in subpaths in reverse order during chained proof verification
+		subpath := []byte(mpath.KeyPaths[len(mpath.KeyPaths)-i-1].String())
+		existProof, ok := p.Proof.(*ics23.CommitmentProof_Exist)
+		if !ok {
+			return sdkerrors.Wrap(ErrInvalidProof, "proof is not an existence proof")
+		}
+		// For subtree verification, we simply calculate the root from the proof and use it to prove
+		// against the value
+		subroot, err := existProof.Exist.Calculate()
+		if err != nil {
+			return sdkerrors.Wrap(ErrInvalidProof, err.Error())
+		}
+		if i != len(proof.Proofs)-1 {
+			if !ics23.VerifyMembership(specs[i], subroot, p, subpath, value) {
+				return sdkerrors.Wrapf(ErrInvalidProof, "invalid proof for path: %s", path.String())
+			}
+		} else {
+			// For the final verification, we prove inclusion against the root that was passed into function
+			// rather than calculating subroot in order to verify that the given value was committed to by
+			// the given root
+			if !ics23.VerifyMembership(specs[i], root.GetHash(), p, subpath, value) {
+				return sdkerrors.Wrapf(ErrInvalidProof, "invalid proof for path: %s", path.String())
+			}
+		}
+		// Each subtree root becomes the proving value for the next proof in the chaining process
+		value = subroot
+	}
+	return nil
 }
 
 // VerifyNonMembership verifies the absence of a merkle proof against the given root and path.
-func (proof MerkleProof) VerifyNonMembership(root exported.Root, path exported.Path) error {
+// VerifyNonMembership verifies a chained proof where the absence of a given path is proven
+// at the lowest subtree and then each subtree's inclusion is proved up to the final root.
+func (proof MerkleProof) VerifyNonMembership(specs []*ics23.ProofSpec, root exported.Root, path exported.Path) error {
 	if proof.IsEmpty() || root == nil || root.IsEmpty() || path == nil || path.IsEmpty() {
-		return errors.New("empty params or proof")
+		return sdkerrors.Wrap(ErrInvalidProof, "empty params or proof")
 	}
 
-	runtime := rootmulti.DefaultProofRuntime()
-	return runtime.VerifyAbsence(proof.Proof, root.GetHash(), path.String())
+	mpath, ok := path.(MerklePath)
```

Comment:
Again you only invite user error when you accept an interface and immediately error if the user didn't pass in one implementation.